### PR TITLE
Problem: memiavl changes are not persisted in real time

### PR DESCRIPTION
**Pull request by pysel**
Closes: #947 

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/wal.go
Comment (yihuang) [inline memiavl/wal.go:67]: I suggest storing the change set of a whole block as a single entry in wal, since that's the atomic unit to replay.

**Comment by yihuang** [comment]
https://github.com/crypto-org-chain/cronos/issues/947 issue created

File: memiavl/wal.go
Comment (pysel) [reply memiavl/wal.go:67]: makes sense! thx

Comment (yihuang) [reply memiavl/wal.go]: For simplicity we can reuse the [change set formats used in versiondb](https://github.com/crypto-org-chain/cronos/blob/bd7251b866bbe3944b997ea002ab846a0dccf6cb/versiondb/client/changeset.go#L31), in memory type would be `[]iavl.KVPair`

Comment (yihuang) [reply memiavl/wal.go]: https://github.com/crypto-org-chain/cronos/pull/952

You can build on top of this API, I've implemented the marshal/unmarshal for the `ChangeSet` type.

Comment (yihuang) [reply memiavl/wal.go]: Another thing is someone need to receive the raw `Set`/`Remove` calls and accumulate the change set, I think we can let the `KVStore` wrapper to do that, so memiavl only expose the `ApplyChangeSet` API, rather than the raw `Set/Remove` APIs.

Comment (pysel) [reply memiavl/wal.go]: oh, I already implemented it but yeah no problem I'll merge your branch, thx for letting me know!

Comment (yihuang) [reply memiavl/wal.go]: I just did a little bit more changes, so now memiavl tree only expose a single method `ApplyChangeSet` to do modifications, I think you can merge the branch now, I just need some review from colleagues to merge.

Comment (yihuang) [reply memiavl/wal.go]: Just realize there's `iavl.ChangeSet` type already, so I changed the PR to reuse that type.

Comment (yihuang) [reply memiavl/wal.go]: #952 is merged.

Comment (pysel) [reply memiavl/wal.go]: applied those changes to this PR

File: memiavl/tree.go
Comment (yihuang) [inline memiavl/tree.go:84]: now you can add the whole change set in `ApplyChangeSet` at once?

Comment (pysel) [reply memiavl/tree.go:84]: yeah, I refactored the code today to use the changeset PR's logic. So basically wal works right now like this: it stores a pair in memory when a `set/remove` is called, and writes to the underlying wal every block's changeset when a `saveVersion` is called.

Right now I am working on `ReplayWAL` function that will basically take the tree at version X, and replay all the changesets stored in a wal until the desired height.

If you have any suggestions on the design, please, lmk!

Comment (yihuang) [reply memiavl/tree.go:84]: I think you can write the change set passed in ApplyChangeSet method directly into WAL, since that's the only entry method now, set/remove won't be called directly by user.

Comment (pysel) [reply memiavl/tree.go:84]: yeah, makes sense, will do that

File: memiavl/tree.go
Comment (yihuang) [inline memiavl/tree.go:84]: Should we just call `t.bwal.writeBlockChanges ` directly here?

File: memiavl/wal.go
Comment (yihuang) [inline memiavl/wal.go:55]: there seems some dead codes.

File: memiavl/tree.go
Comment (pysel) [reply memiavl/tree.go:84]: `Flush()` does that internally. but I see what you mean. I think I can just get rid of `BlockChangeset` in wal struct, because we do not need it anymore, since `set/remove` are not exposed

File: memiavl/tree.go
Comment (yihuang) [inline memiavl/tree.go:167]: - I think we can call `ApplyChangeSet` here directly, so don't need to decode all the change sets in memory.
- we don't have to update hash for the intermediate versions, we can do `tree.ApplyChangeSet(, false)`, and only do a `tree.RootHash()` at the end.

File: memiavl/tree.go
Comment (yihuang) [reply memiavl/tree.go:84]: yeah, that'd be simpler.

Comment (pysel) [reply memiavl/tree.go]: ok, changed that, now changesets can be directly provided as `Flush` args

Comment (pysel) [reply memiavl/tree.go]: I will also move `Flush` to be ran before all sets and removes

File: memiavl/tree.go
Comment (pysel) [reply memiavl/tree.go:166]: 1) If we do that, if a user provides a version that does not exist in wal, the changes to the tree will be applied up to the version that exists in wal and then error. Is that desired?

Comment (pysel) [reply memiavl/tree.go:166]: I also will add an option to simply sync all versions available in the wal

Comment (pysel) [reply memiavl/tree.go:175]: added a way to sync untill the end of wal: if `untilVersion` is 0, it will sync till the end

Comment (pysel) [reply memiavl/tree.go:175]: about the error still seems weird to me, it would mean that the tree is corrupted, but why would we declare it corrupted if that can be prevented? But the final decision is up to you

Comment (pysel) [reply memiavl/tree.go:175]: also want to add function to read batches of changesets from wal in one go, will improve readability in `ReplayWAL` I believe

Comment (pysel) [reply memiavl/tree.go:175]: nvm, I am using the native `wal.Read()` in `ReplayWAL`, not the wrapper, but I will propose batch reads to tidwall/wal

Comment (yihuang) [reply memiavl/tree.go:175]: > about the error still seems weird to me, it would mean that the tree is corrupted, but why would we declare it corrupted if that can be prevented? But the final decision is up to you

just want to lower the memory usage here, it could be pretty big depending on the snapshot interval.

**Review by yihuang — APPROVED** [review]
LGTM, thanks a lot for the contribution, I'll probably do some adjustments when doing the final integration.

**Comment by yihuang** [comment]
@pysel can you run `go mod tidy && gomod2nix` in root directory, or allow me to push the changes to the branch?

File: memiavl/tree.go
Comment (pysel) [reply memiavl/tree.go:175]: yeah your point makes sense. What do you think about copying the tree at the beginning of the function and applying changes to it the way you suggested? still a memory overhead, but not that big I guess

Comment (yihuang) [reply memiavl/tree.go:175]: I think if anything unexpected happens with WAL, it's almost always a serious unrecoverable issue.

Comment (pysel) [reply memiavl/tree.go:175]: oh, I see what you mean. What I meant here is a user's mistake, aka providing an `untilVersion` that is higher than the latest version from the wal, which could result in `wal.Read()` error. I will just add a check that the version provided is within boundaries and we should be good to go with your design, thanks!

**Review by yihuang — CHANGES_REQUESTED** [review]
The wal Flush seems always fail for "out of order" error

File: memiavl/tree.go
Comment (mmsqe) [inline memiavl/tree.go:90]: yes, pointer was wrong [use](https://medium.com/swlh/use-pointer-of-for-range-loop-variable-in-go-3d3481f7ffc9), that var will be reused in for loop

**Comment by pysel** [comment]
> The wal Flush seems always fail for "out of order" error in unit test.

checking

**Comment by pysel** [comment]
@yihuang ah, the reason is that the wal file is not deleted at the end of some tests, so it tries to write indices to the file that already exist there since the last version, let me fix it

**Review by yihuang — APPROVED** [review]

File: memiavl/tree_test.go
Comment (mmsqe) [inline memiavl/tree_test.go:207]: might need mv the Cleanup before each test

Comment (pysel) [reply memiavl/tree_test.go:203]: these subtests do not create any new wals, so I do not think we should

**Review by mmsqe — APPROVED** [review]

**Pull request merged**
